pennylane/circuit_graph.py

```diff
@@ -12,305 +12,278 @@
 
                 # check if any of the dependents are in the
-                # existing layer
-                if set(layer_ops[layer][0]) & sub:
-                    # operation depends on previous layer,
-                    # start a new layer count
-                    layer += 1
+                # currently assembled layer
+                if set(current.ops) & sub:
+                    # operation depends on current layer, start a new layer
+                    current = Layer([], [])
+                    layers.append(current)
 
                 # store the parameters and ops indices for the layer
-                layer_ops.setdefault(layer, ([], []))
-                layer_ops[layer][0].append(op_idx)
-                layer_ops[layer][1].append(param_idx)
+                current.ops.append(op)
+                current.param_inds.append(param_idx)
 
-        return [Layer(*k) for _, k in sorted(list(layer_ops.items()))]
+        return layers
 
     def iterate_layers(self):
-        """Identifies and returns an iterable containing
-        the parametrized layers.
+        """Parametrized layers of the circuit.
 
         Returns:
             Iterable[tuple[list, list, tuple, list]]: an iterable that returns a tuple
             ``(pre_queue, layer, param_idx, post_queue)`` at each iteration.
 
             * ``pre_queue`` (*list[Operation]*): all operations that precede the layer
 
-            * ``layer`` (*list[Operation]*): the parametrized gates in the layer
+            * ``layer`` (*list[Operation]*): parametrized gates in the layer
 
-            * ``param_idx`` (*tuple[int]*): The integer indices corresponding
-              to the free parameters of this layer, in the order they appear in
-              this layer.
+            * ``param_inds`` (*tuple[int]*): integer indices corresponding
+              to the free parameters of this layer, in the order they appear
 
             * ``post_queue`` (*list[Operation, Observable]*): all operations that succeed the layer
         """
         # iterate through each layer
-        for ops, param_idx in self.layers:
-
-            # get the ops in this layer
-            layer = self.get_ops(ops)
-            pre_queue = self.get_ops(self.ancestors(ops))
-            post_queue = self.get_ops(self.descendants(ops))
+        for ops, param_inds in self.layers:
+            pre_queue = self.ancestors_in_order(ops)
+            post_queue = self.descendants_in_order(ops)
+            yield pre_queue, ops, tuple(param_inds), post_queue
 
-            yield pre_queue, layer, tuple(param_idx), post_queue
+    def update_node(self, old, new):
+        """Replaces the given circuit graph node with a new one.
 
-    def update_node(self, node, op):
-        """
-        Updates a given node with a new operation, op.
+        .. note:: Does alter the graph edges in any way. variable_deps is not changed, _grid is not changed.
+           Dangerous, do we need this?
```

Comment:
iirc the ability to temporarily replace operations is required by some of the parameter shift rule methods. Is it possible to remove this/avoid this?

Reply:
Looks like only `QNode._pd_analytic_var` uses `update_node`. During the next step in the refactoring I'll check if it's required.

Reply:
What was the outcome of this? I had a conversation to @zzzeid, and he mentioned that `QNode._pd_analytic_var` doesn't work without the safe node update.

Reply:
(let's move the note to a comment TODO)

Reply:
`update_node` is ok as long as you remember to undo the change once you're done. I changed it into a TODO.